proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,439 @@
+
+As a result, there are multiple cryptography libraries with similar interfaces, and it is not always clear which should be used preferentially. This promotes rot as downstream users unknownly rely on unmaintained libraries that slowly stop working.
+
+Furthermore, a number of important libraries in the cryptography ecosystem (eg, `cryptonite`, `asn1-encoding`, and `memory`) have recently been archived by their author and will no longer be receiving updates. Their backend is written in C, which presents a long-term maintenance burden with a high knowledge and skill requirement. NaCl bindings such as `saltine` are a high-quality alternative, but solve only a limited set of problems.
+
+The resulting brittleness has placed the long-term stability of important libraries such as `crypton`, `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented. Additionally the instability creates an unattractive environment for developers, effectively driving them away to other ecosystems that can provide the necessary stability.
+
+There are several additional reasons for this lack of suitable engineers:
+
+1) Cryptography involves a lot of stateful low-level bitwise manipulation and non-determinism, something Haskell is not known for being good at. Classes like `Bits` and `FiniteBits` classes are critical for cryptography, but are terribly awkward to implement (like `Num`). Notably, `ByteString` does not implement them for nuanced reasons (see Appendix B - Lack of ByteString `Bits` instance).
+
+2) Cryptography has a high knowledge and technical skill floor. Cryptography is more than just the usual primitives of ciphers, public and private keys, hashes, and MACs, and the scope of knowledge can grow quickly as primitives are combined into more advanced algorithms.
+
+3) There is a persistent notion that functional languages are too slow for cryptography, despite it being based on historical reasons that have been mostly since addressed. Well-written modern Haskell approaches C in terms of speed, and has a much higher ergonomic-use and type-safety factor, but the public image has not caught up yet, and so the ecosystem has not received the investment and attention necessary to attract the attention of skilled engineers for developing either pure implementations or foreign bindings.
+
+4) A fair number of cryptographic projects in recent years has been specifically towards *cryptocurrencies*, rather than *cryptography in general*, and their results are at large fairly useless if they cannot be used outside of their ecosystems.
```

Comment:
Maybe this could be rephrased slightly.

I.e. I think _some_ of the cryptocurrency work has probably usefully affected some libraries (maybe?); it's just the scope of a lot of crypto work in cryptocurrencies can be idiosyncratic and not generally useful in scenario A, B, C ...

This could go hand-in-hand with a change to this section (above):

> NaCl bindings such as `saltine` are a high-quality alternative, but solve only a limited set of problems.

to list the problems that saltine focuses on (I actually don't know!). For example, personally in my daily life I work with JWTs and Cookies; both of which require cryptography. I recently learned (from @Kleidukos ) that the clientsession library depends on cryptonite (likewise [jose](https://flora.pm/packages/%40hackage/jose/0.10/dependencies). So for me, I only need _these_ libraries to improve, not the lower-level ones.

I.e. it would help me to understand a bit more what is the list of things that I can't do with saltine but can do with botan.